torchdynamo/optimizations/distributed.py

```diff
@@ -0,0 +1,145 @@
+import torch.fx as fx
+from typing import List
+import torch
+from contextlib import contextmanager
+
+class GraphProducer(object):
+    def __init__(self):
+      self.possible_input_names: set[str] = set()
+      self.inputs: list[str] = []  # this is used to iterate the above set given it's iteration is abitrary.
+      self.code: list[fx.Node] = []
+      self.possible_output_names: set[str] = set()
+      self.outputs: list[fx.Node] = []
+      self._graph: fx.Graph = None
+
+    def graph(self):
+      if self._graph is not None:
+        return self._graph
+
+      self._graph = fx.Graph()
+      for input in self.inputs:
+        self._graph.placeholder(input)
+      for code in self.code:
+        self._graph.node_copy(code)
+      self._graph.output(tuple(self.outputs))
+      return self._graph
+
+class DDPOptimizer:
+    def __init__(self, bucket_cap_mb: int, parameters_to_ignore: List[str], backend_compile_fn):
+        self.bucket_cap_mb = bucket_cap_mb
+        self.parameters_to_ignore = parameters_to_ignore
+        self.backend_compile_fn = backend_compile_fn
+    
+    def compile_fn(self, gm: fx.GraphModule, example_inputs: List[torch.Tensor]):
+        """
+        TODO:
+        - handle params_and_buffers_to_ignore
+        - implement call_method/call_forward
+        - handle kwargs
+        """
+
+        print("DDPOptimizer called with FX graph:")
+        gm.graph.print_tabular()
+        print()
+        # 1. Splitting the gm into small graphs, following DDP parameter bucketing
```

Comment:
That means this logic needs to be synchronized with the DDP bucketing logic, is that right

Reply:
right. 
1) ddp first buckets based on reverse order of parameter declaration in nn module structure
2) then after first run, ddp rebuckets (incurs tensor copies) using the actual execution order of first backward, which can be different
3) ddp's actual bucketing logic is pretty simple, given an order of parameters and a set of parameters to exclude, it fills buckets to capacity